### PR TITLE
Enable throttling apigatway at api, stage and method levels

### PR DESCRIPTION
**Pull request by RossMurr4y**
## Description

Implements hamlet-io/engine#1376 on the `apigateway` component. This is server-side throttling to protect the backend - it is not client-side throttling.

* Throttling defined at 3 independent levels: the whole API (aka defaults), the stage (overwriting any defaults) and individual methods (overwriting any of the others that would have applied to those methods)
* Throttling can be defined at two levels:
  * the stage
  * On a pattern, which implements the throttling on matching methods

Example `apigw.json` shows the following:

* stage level throttling is set to a burst of 5000, and a rate of 800
* Any methods patching the regex path `/.*/search` with a verb of `post` will be throttled to a burst of 1000 and rate of 300.
**Numbers are for example only, they are not appropriate throttle figures.**
```json
{
    "apigw" : {
        "Internal" : true,
        "Value" : {
            "Type" : "lambda",
            "Variable" : "IMP_API_LAMBDA",
            "BinaryTypes" : ["*/*"],
            "Throttling": {
                "BurstLimit" : 5000,
                "RateLimit" : 800
            },
            "Patterns" : [
                {
                    "Path" : "/.*/search",
                    "Verb" : "post",
                    "Variable" : "FLIBBERDY",
                    "BurstLimit" : 1000,
                    "RateLimit" : 300
                }
            ]
        }
    }
}
```

## Motivation and Context

Would like to be able to control the rate at which the API handles incoming requests to prevent overwhelming the backend in some scenarios.

## How Has This Been Tested?

Local template generation and deployment.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves the structure or operation of the implementation)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Followup Actions

- [x] None

## Checklist:

- [ ] My change requires a change to the [documentation](https://github.com/hamlet-io/docs).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] None of the above.

### CODE REVIEW

**Comment by RossMurr4y** [comment]
Checks will of course fail until hamlet-io/engine#1376 is merged

File: aws/components/apigateway/state.ftl [hunk 1]
```diff
@@ -83,6 +83,7 @@ created in either case.
 
     [#local certificatePresent = isPresent(solution.Certificate) ]
     [#local cfPresent          = isPresent(solution.CloudFront) ]
+    [#local cfPresent          = isPresent(solution.CloudFront) ]
```
Comment (ml019) [inline aws/components/apigateway/state.ftl:86]: Looks redundant?

File: aws/components/apigateway/setup.ftl [hunk 2]
```diff
@@ -303,21 +303,57 @@
...
                     "RestApiId" : getReference(apiId),
-                    "MethodSettings": [
-                        {
-                            "HttpMethod": "*",
-                            "ResourcePath": "/*",
-                            "LoggingLevel": "INFO",
-                            "DataTraceEnabled": true
-                        }
-                    ],
+                    "MethodSettings": methodSettings,
```
Comment (ml019) [inline aws/components/apigateway/setup.ftl:356]: We only want to set the defaults if there is throttling settings defined. so I think we want this to be added via `AttributeIfContent()`

File: aws/components/apigateway/setup.ftl [hunk 3]
```diff
@@ -303,21 +303,57 @@
...
+
+        [#-- Throttling Configuration --]
+        [#local openapiIntegrations = 
+            getOccurrenceSettingValue(occurrence, [["apigw"], ["Integrations"]], true)]
+
+        [#local methodSettings = [
+            {
+                "HttpMethod": "*",
+                "ResourcePath": "/*",
+                "LoggingLevel": "INFO",
```
Comment (ml019) [inline aws/components/apigateway/setup.ftl:315]: Don't want this I think unless explicitly defined - unless mandatory

File: aws/components/apigateway/setup.ftl [hunk 4]
```diff
@@ -303,21 +303,57 @@
...
+                "DataTraceEnabled": true
```
Comment (ml019) [inline aws/components/apigateway/setup.ftl:316]: Don't want this I think unless explicitly defined - unless mandatory

File: aws/components/apigateway/setup.ftl [hunk 5]
```diff
@@ -303,21 +303,57 @@
...
+                openapiIntegrations.Throttling.RateLimit)
+        ]]
+
+        [#-- Integration Patterns (as Regex) into Matching Method Throttling (as explicit paths) --]
+        [#if definitionsObject[core.Id]?? ]
+            [#list definitionsObject[core.Id].paths as path,pathConfig]
+                [#list pathConfig?keys as verb]
+                    [#list openapiIntegrations.Patterns as pattern]
+                        [#if path?matches(pattern.Path)] 
+                            [#local methodSettings = combineEntities(
```
Comment (ml019) [inline aws/components/apigateway/setup.ftl:332]: can just do array addition - I think it would be clearer.

**Review by ml019 — CHANGES_REQUESTED** [review]

File: aws/components/apigateway/state.ftl (hunk 1 above)
Comment (RossMurr4y) [reply aws/components/apigateway/state.ftl]: Yep, Not sure how that was added in there. Removed now.

File: aws/components/apigateway/setup.ftl (hunk 2 above)
Comment (RossMurr4y) [reply aws/components/apigateway/setup.ftl]: Yep certainly should be - fixed up

File: aws/components/apigateway/setup.ftl (hunk 3 above)
Comment (RossMurr4y) [reply aws/components/apigateway/setup.ftl]: Amended with a wrapper if-statement, and instantiating an empty array.

File: aws/components/apigateway/setup.ftl (hunk 5 above)
Comment (RossMurr4y) [reply aws/components/apigateway/setup.ftl]: Updated

File: aws/components/apigateway/setup.ftl [hunk 6]
```diff
@@ -303,27 +303,64 @@
...
+                    openapiIntegrations.Throttling.RateLimit)
+            ]]
+        [/#if]
+
+        [#-- Integration Patterns (as Regex) into Matching Method Throttling (as explicit paths) --]
+        [#if definitionsObject[core.Id]?? ]
+            [#list definitionsObject[core.Id].paths as path,pathConfig]
+                [#list pathConfig?keys as verb]
+                    [#list openapiIntegrations.Patterns as pattern]
+                        [#if path?matches(pattern.Path)]
```
Comment (ml019) [inline aws/components/apigateway/setup.ftl:334]: Needs verb comparison as well

File: aws/components/apigateway/setup.ftl [hunk 7]
```diff
@@ -303,27 +303,64 @@
...
+            ]]
+        [/#if]
+
+        [#-- Integration Patterns (as Regex) into Matching Method Throttling (as explicit paths) --]
+        [#if definitionsObject[core.Id]?? ]
+            [#list definitionsObject[core.Id].paths as path,pathConfig]
+                [#list pathConfig?keys as verb]
+                    [#list openapiIntegrations.Patterns as pattern]
+                        [#if path?matches(pattern.Path)] 
+                            [#local methodSettings +=
```
Comment (ml019) [inline aws/components/apigateway/setup.ftl:335]: Should only generate entry if the throttling element is defined in the pattern.

File: aws/components/apigateway/setup.ftl [hunk 8]
```diff
@@ -303,27 +303,64 @@
             outputs={}
             dependencies=apiId
         /]
+
+        [#-- Throttling Configuration --]
+        [#local methodSettings = []]
+        [#local openapiIntegrations = 
+            getOccurrenceSettingValue(occurrence, [["apigw"], ["Integrations"]], true)]
+
+        [#if openapiIntegrations.Throttling?has_content]
```
Comment (ml019) [inline aws/components/apigateway/setup.ftl:312]: My bad. Looking at original code, we are always creating at least one methodSetting entry so this code doesn't need to be wrapped in the check for throttling.

File: aws/components/apigateway/setup.ftl [hunk 9]
```diff
@@ -303,27 +303,64 @@
...
-                            "DataTraceEnabled": true
-                        }
-                    ],
                     "StageName" : stageName,
                     "AccessLogSetting" : {
                         "DestinationArn" : getArn(accessLgId),
                         "Format" : "$context.identity.sourceIp $context.identity.caller $context.identity.user $context.identity.userArn [$context.requestTime] $context.apiId $context.httpMethod $context.resourcePath $context.protocol $context.status $context.responseLength $context.requestId"
                     }
                 } +
+                attributeIfContent("MethodSettings", methodSettings) +
```
Comment (ml019) [inline aws/components/apigateway/setup.ftl:363]: Again, my bad. We always want this attribute but I think we leave the code as you now have it. It works the same and would permit a case where we didn't want to always provide method defaults.

**Review by ml019 — CHANGES_REQUESTED** [review]

File: aws/components/apigateway/setup.ftl (hunk 8 above)
Comment (RossMurr4y) [reply aws/components/apigateway/setup.ftl]: Updated: 15552a5

File: aws/components/apigateway/setup.ftl (hunk 6 above)
Comment (RossMurr4y) [reply aws/components/apigateway/setup.ftl]: Updated: 8f2a517

File: aws/components/apigateway/setup.ftl (hunk 7 above)
Comment (RossMurr4y) [reply aws/components/apigateway/setup.ftl:333]: Updated: 464cde1

**Pull request merged**